feat: Donation request device message (#6913)

iequidoo opened the pull request:
See commit messages. There are no tests (the change is as simple as possible, i don't want to maintain tests for such "features"), have done a quick check in Desktop.
Close #6913

r10s reviewed (commented):
thanks a lot for pushing that forward!

i find review a bit hard because of non-telling names, what about using sth. as 
`days_until_first_donation_request`, `days_until_next_donation_request` and `LastDonationRequestTimestamp` (instead of _period_check, period_msg, DonationRequestTs_)?

also, not sure if it is needed to check on every sending of a message - even though the expensive check seems to be called only if clock is later than `LastDonationRequestTimestamp`. but maybe that's fine, no blocker.

for the expensive sql command: the mentioned corner case in the issue is about shared-usage or classic-usage, if ppl are doing that they seem to be into delta chat as well, as we do not download history. if would keep it simple here and look at all messages

also, i would add the message only if there are more messages, eg. 100, 10 messages is not really real usage.

finally, the stock string is missing in the deltachat.h

iequidoo commented:
> `days_until_first_donation_request`, `days_until_next_donation_request` and `LastDonationRequestTimestamp` (instead of _period_check, period_msg, DonationRequestTs_)?

Renamed to `secs_between_checks` (it's only a consequence that the first request appears after this time), `secs_between_requests`, `DonationRequestNextCheck` (we don't always add `Timestamp`, e.g. see `LastHousekeeping`).

> also, not sure if it is needed to check on every sending of a message

We have the config cache, so looking at `Config::DonationRequestNextCheck` even on every sending is fine.

> for the expensive sql command: the mentioned corner case in the issue is about shared-usage or classic-usage, if ppl are doing that they seem to be into delta chat as well, as we do not download history. if would keep it simple here and look at all messages

Removed the check for whether a message is encrypted. Now it should be fast because we have `msgs_index7 ON msgs (state, hidden, chat_id)`. Anyway i'm going to add a new `OutRcvd` message state in #5813 (it's needed to fix the remaining message sorting issues that e.g. i have with Gmail) that will fix corner cases, i.e. we can only count locally sent messages.

> also, i would add the message only if there are more messages, eg. 100, 10 messages is not really real usage.

Made it 50. 100 looks too much for me personally, maybe in average people chat more, idk.

> finally, the stock string is missing in the deltachat.h

Fixed.

r10s commented:
after some internal discussions with @adbenitez and others:

- for now, we do not want the message to re-appear every some time, we can add that at some point, but initially it is not wanted
- instead, the 14 days should be at least 4 weeks, and the number of messages that should be there should be more, say, 100
- the wording "X cost time and time cost money" should be shortened to "X cost money" - ppl do not read much, so no need to be too nerdy  and pretentious here. ppl also do not say "bread does not cost money. making a bread cost time, and time may cost money" ... 
- "seems you're using Delta Chat" was indeed a bad suggestion from my side, what @adbenitez suggested in the issue is better, "❤️ Seems you're enjoying Delta Chat!"
the final wording is anyway in android, but it seems good to have a not-so-bad-fallback

r10s reviewed (approved):
yeah, this lgtm now, also tested once :)
